### PR TITLE
move authorization out of the authentication plugin

### PR DESCRIPTION
**Pull request by Geal**
*Breaking change*: this changes the JWT plugin's behaviour, it now lets all requests go through

now the authentication plugin only rejects invalid tokens (bad signature, expired, etc) but lets requests without tokens go through. The authorization decision is then handled later.
As an example, a rhai script can be used to refuse all unauthenticated requests

This also adds support for the `extensions` field in GraphQL errors generated from rhai

TODO:
- [ ] docs
- [ ] warn existing users about the change

*Description here*

Fixes #issue_number

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: apollo-router/src/plugins/rhai/mod.rs [hunk 1]
```diff
@@ -1131,40 +1137,40 @@ fn process_error(error: Box<EvalAltResult>) -> ErrorDetails {
...
-                        ed_message = Some(value);
-                    }
-                }
-
-                if let Some(status) = ed_status {
+                if let Some(status) = map
+                    .get("status")
+                    .and_then(|status_dyn| status_dyn.as_int().ok())
+                    .and_then(|value| StatusCode::try_from(value as u16).ok())
+                {
```
Comment (garypen) [inline apollo-router/src/plugins/rhai/mod.rs:1144]: I like this. I should always go back and check for useful combinators when I finish writing code...

File: apollo-router/tests/fixtures/require_authentication.rhai [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    service.map_request(request_callback);
+}
+
+// This will examine our context to determine if our user is authenticated.
+fn process_request(request) {
+    let claims = request.context[Router.APOLLO_AUTHENTICATION_JWT_CLAIMS];
+    if claims == () {
+        throw #{
+            status: 401,
+            message: "Missing 'Authorization' header",
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/tests/fixtures/require_authentication.rhai:12]: right, that's better. That script is only there for the tests, I have not yet written the example and docs

[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/tests/fixtures/require_authentication.rhai]: changed in [a73f9af](https://github.com/apollographql/router/pull/2708/commits/a73f9af7351f948e9030373bdc81e3374146bb52)

File: apollo-router/src/plugins/authentication/tests.rs [hunk 3]
```diff
@@ -138,7 +166,7 @@ async fn it_rejects_when_there_is_no_auth_header() {
     .unwrap();
 
     let expected_error = graphql::Error::builder()
-        .message("Missing authorization header")
+        .message("Missing 'Authorization' header")
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/plugins/authentication/tests.rs]: changed in [a73f9af](https://github.com/apollographql/router/pull/2708/commits/a73f9af7351f948e9030373bdc81e3374146bb52)

File: apollo-router/src/plugins/rhai/mod.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/plugins/rhai/mod.rs:1145]: it can take a while to write with the combinators, but then it's a bit nicer to follow what's going on

**Review by garypen — APPROVED** [review]
Looks good!

**Pull request merged**